### ISSUE TITLE
Add role=form to generated Form::open()

### ISSUE THREAD

**Issue by ayumihamsaki** [octobercms/october#4253]
Not sure whether to file this as a bug or feature request?

## The issue:

The following October info web page: https://octobercms.com/docs/markup/function-form

We can add the php code snippet:

```php
<?= Form::open() ?>
```

This produces something like this:

```html
<form method="POST" action="https://www.example.com/something" accept-charset="UTF-8"><input name="_session_key" type="hidden" value="uFgJgbwv1G5435srou6q4fGrpXzLlMoKtK9spP">
<input name="_token" type="hidden" value="mOg99PjpR9SjfSyw6uZP543XCyPsJx0ayb">
```

The trouble is that the output doesn't add any accessibility standards.

## Solution

By default October CMS should add the following snippet:

```html
role="form"
```

Therefore giving a standard output of this:

```html
<form method="POST" action="https://www.example.com/something" accept-charset="UTF-8" role="form"><input name="_session_key" type="hidden" value="uFgJgbwv1G5435srou6q4fGrpXzLlMoKtK9spP">
<input name="_token" type="hidden" value="mOg99PjpR9SjfSyw6uZP543XCyPsJx0ayb">
```

This may not sound a big deal, but this makes things very hard for a developer to want to do things properly with regards to accessibility standards.

## Performance Wise

Adding `role="form"` to the code as default adds only benefits and no side effects to any website! It should be injected into the website as standard.

**Comment by LukeTowers** [comment]
@ayumihamsaki feel free to submit a PR to accomplish that.

**Comment by ayumihamsaki** [comment]
@LukeTowers I would be happy to do that, but I can't find the correct file to update. Could you point me in the right direction please.

**Comment by LukeTowers** [comment]
@ayumihamsaki it might be something in the library (Support perhaps? Html?) or in Laravel itself

**Comment by w20k** [comment]
@ayumihamsaki should be here -> `vendor\october\rain\src\Html\FormBuilder: 112-160, 158 where attributes are converted to a string` 👍

**Comment by ayumihamsaki** [comment]
@w20k Thanks :+1: 

https://github.com/octobercms/library/blob/master/src/Html/FormBuilder.php

**Comment by bezgodov** [comment]
@ayumihamsaki I've just finished developing website on October and checked on https://validator.w3.org and got a warning The form role is unnecessary for element form

As it's written on http://w3c.github.io/html-aria/#rules-wd
Web developers SHOULD NOT set the ARIA role and aria-* attributes to values that match the implicit ARIA semantics defined in the table.

And additional info you may read here: https://stackoverflow.com/a/35402845

I'm not very good at accessibility, but it intuitively seems strange to me when we use <form> tag within role="form"

Don't I miss something? (I haven't tested but suppose that all screen readers understand that it's a form by <from> tag [and there is no need in role="form"]). Different case when we use <div> as a form and of course it's useful to do so <div role="form">

Let me know if I am wrong in something

P.S. In my opinion, I suggest it's better to leave possibility to add role of any kind to form tag (form_open ({ 'role': 'yourRole'}) as it is at the moment, but not by default.

By the way I didn't find an option to remove role tag from form using form_open()

**Comment by ayumihamsaki** [comment]
@bezgodov The problem is that w3c made the rule to add role attributes to html-aria landmarks and then changed the rules a while ago. It's there for backwards compatibility for the screen readers. Having it there will not cause any issues at all. If that validator is flagging it as an error then they need to update it to a warning message.

**Comment by ayumihamsaki** [comment]
@bezgodov I can give you the answer "backwards compatibility for screen readers"

The old method was to add `role="form"` even to the `<form>` tags and the reasson was:

> If you add a `role="form"` to a form, a screen reader sees it as a region on a webpage. That means that a user can easily jump to the form with his/her region quick navigation keys (for example, in JAWS 15 you use R for this). And also, your user will be able to easily find where the form starts and ends because screen readers mark start and end of regions.

Then W3c changed the spec to say about HTML5 land marks elements such as `<table>` and `<form>` etc. that you don't need to add `role="form"` into `<form>` anymore.

However, a person using a screen reader treats it like an extension of their body. It was discussed a while back whether to not support backwards compatibility for screen readers or not? The conclusion was to support backwards compatibility.

A famous example I can give is with Stephen Hawking, they invented an upgraded voice for him that sounded more human but he didn't want to upgrade to the new system as he said that the old robotic voice was how he saw himself sounded.

People with screen readers don't like to update all the time as they get use to how that thing works and stay with that version. Many people use a screen reader version that is quite old. One screen reader person said they perfer the older version as it's less complicated and didn't want to update etc.

Basically that's the overview from that conversation a while back with regards to supporting backwards compatibility.

It won't hurt your website in anyway having `role="form"` and allows 100% of all screen readers to work on your website.

I leave it for the admins to decide anyway...

**Comment by bezgodov** [comment]
@ayumihamsaki thank you for you reply, I just care about warning on http://validator.w3.org, cause there are some technical reports which contain something like this `after deploy test website for any errors, warnings on http://validator.w3.org and get 100 points on Google Speed Insight (something like this)` and in our case I won't be able to do this and explain why (what is much harder than it should be)

I got you point and basically agree with you, thank you one more time

**Comment by LukeTowers** [comment]
@ayumihamsaki what does role="form" actually do on the older screenreaders?

**Comment by ayumihamsaki** [comment]
@bezgodov Google Speed Insight doesn't help much with performance, I would recommend you look at Lighthouse instead.

**Comment by ayumihamsaki** [comment]
@LukeTowers 

> If you add a role="form" to a form, a screen reader sees it as a region on a webpage. That means that a user can easily jump to the form with his/her region quick navigation keys (for example, in JAWS 15 you use R for this). And also, your user will be able to easily find where the form starts and ends because screen readers mark start and end of regions.

So basically what that means is instead of a user having to scroll through every single element on the web page. They can jump the web page to the form. Other role landmarks that W3c says not to use are things like `<article>` and `role="article"` would allow you to jump from the top of the web page and start reading the Primary article with a single click - instead of having to scroll through the top navigation menu etc.

Older screen readers used role and not HTML5. New screen readers use the HTML5.

Hope that explains it.

For me personally I have several goals for web pages and one is conversions, so allowing users to get to the main content or checkout forms in the least amounts of clicks is always a good thing.

Timeline of W3c change is roughly:

5 years ago it was correct to use role in landmark elements.

3 years ago w3c changed the rules.

However, many people still use very old screen readers and really dont want to change or upgrade. So even though it's 5 years it's not like how it is for regular people. These screen readers are really personal to those people.

I leave it up to you if you want to remove it or not? By the way there is two sets of PR's with this. I did one pr for the frontend and one for the backend. Just in case people did complain.

**Comment by ayumihamsaki** [comment]
@LukeTowers 

> do you know of any specific screenreaders / screenreader versions that actually relied on role="form" on elements?

Yah there are quite a few screen readers that replied on using that old method, as it was the official method a few years ago.

> It feels like this could be something that was just not properly specified in the standards but would have been common practice to just treat form elements as form regions.

I give you two form examples with different meanings to explain the logic behind the original idea:

### Example 1 - Contact form (old method)

```html
<form role="form">
  <input type="text" name="firstname" value="Mickey">
  <input type="text" name="lastname" value="Mouse">
  <input type="submit" value="Submit">
</form>
```

### Example 2 - Search form (current method)

```html
<form role="search">
  <input type="search" aria-label="search text" size="20">
  <input type="submit" value="Search">
</form>
```

To a screen reader both are the same and have the HTML5 `<form>` yet the `role` attribute helps the screen reader to understand it better.

Now the new method to adding `role="form"` is like this:

### Example 3 - Contact form (new method)

```html
  <div role="form" aria-labelledby="contact">
    <fieldset>

      <legend id="contact">Add Contact </legend>

    ... form controls add contact ...

    </fieldset>
  </div>
```

They have basically said if the HTML semantic markup is using `<form>` don't use `role` and if the  HTML semantic markup is using something else like `<div>` then use the attribute `role="form`".

---

If you want I can revert the frontend one and keep the backend ones. I believe @bezgodov example is referrencing the frontend. I can't image people wanting to validate the backend.

Just be mindful this is a `warning` and not an `error` - so W3c is kind of saying it as a recommendation, for people to move to the updated spec, but knows it can not say this as an `error` due to backwards compatibility issues.

**Comment by ayumihamsaki** [comment]
@LukeTowers 

> do you know of any specific screenreaders / screenreader versions that actually relied on role="form" on elements?

Yah there are quite a few screen readers that replied on using that old method, as it was the official method a few years ago.

> It feels like this could be something that was just not properly specified in the standards but would have been common practice to just treat form elements as form regions.

I give you two form examples with different meanings to explain the logic behind the original idea:

### Example 1 - Contact form (old method)

```html
<form role="form">
  <input type="text" name="firstname" value="Mickey">
  <input type="text" name="lastname" value="Mouse">
  <input type="submit" value="Submit">
</form>
```

### Example 2 - Search form (current method)

```html
<form role="search">
  <input type="search" aria-label="search text" size="20">
  <input type="submit" value="Search">
</form>
```

To a screen reader both are the same and have the HTML5 `<form>` yet the `role` attribute helps the screen reader to understand it better.

Now the new method to adding `role="form"` is like this:

### Example 3 - Contact form (new method)

```html
  <div role="form" aria-labelledby="contact">
    <fieldset>

      <legend id="contact">Add Contact </legend>

    ... form controls add contact ...

    </fieldset>
  </div>
```

They have basically said if the HTML semantic markup is using `<form>` don't use `role` and if the  HTML semantic markup is using something else like `<div>` then use the attribute `role="form"`.

---

If you want I can revert the frontend one and keep the backend ones. I believe @bezgodov example is referrencing the frontend. I can't image people wanting to validate the backend.

Just be mindful this is a `warning` and not an `error` - so W3c is kind of saying it as a recommendation, for people to move to the updated spec, but knows it can not say this as an `error` due to backwards compatibility issues.

---

Basically W3c has converted it to shorthand code.

### Old screen readers

- `<form>` - no keyboard shortcut method as we dont know what the role of this form is?
- `<form role="search">` - can do a shortcut method as we know it's a search form.
- `<form role="form">` - can do a shortcut method as we know it's a standard form.

### New screen readers

- `<form>` - can do a shortcut method as we know it's a standard form.
- `<form role="search">` - can do a shortcut method as we know it's a search form.
- `<form role="form">` - can do a shortcut method as we know it's a standard form.

**Comment by ayumihamsaki** [comment]
For ref: https://github.com/octobercms/library/blob/cabdf96a621db360e09049c7cb94b98c72d62b4b/src/Html/FormBuilder.php#L139
